paddle/operators/rank_loss_op.cc

```diff
@@ -0,0 +1,105 @@
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    // input check
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("P"), "Input(P) shouldn't be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Oi"), "Input(Oi) shouldn't be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Oj"), "Input(Oj) shouldn't be null");
+    auto p_dims = ctx.Input<framework::Tensor>("P")->dims();
+    auto oi_dims = ctx.Input<framework::Tensor>("Oi")->dims();
+    auto oj_dims = ctx.Input<framework::Tensor>("Oj")->dims();
+    PADDLE_ENFORCE_EQ(oi_dims, oj_dims,
+                      "Input(Oi) and Input(Oj) must have the same size");
+    PADDLE_ENFORCE_EQ(
+        p_dims, oi_dims,
+        "Input(P) must have the same size with Input(Oi) & Input(Oj)");
+    ctx.Output<framework::Tensor>("Out")->Resize(p_dims);
+  }
+};
+
+class RankLossOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  RankLossOpMaker(framework::OpProto *proto,
+                  framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("P", "The desired target values for posteriors.");
+    AddInput("Oi", "The model output for item i.");
+    AddInput("Oj", "The model output for item j.");
+    AddOutput("Out", "The output tensor of RankLoss operator.");
+    AddComment(R"DOC(RankLoss operator
+
+A rank loss operator for learning to rank (LTR) task. This operator contains
+three inputs: P, Oi, and Oj, and the rank cost can be expressed as
+
+\f[
+  C_{i,j} = -\tilde{P_{ij}} * o_{i,j} + log(1 + e^{o_{i,j}}) \\
+  o_{i,j} =  o_i - o_j  \\
+  \tilde{P_{i,j}} = \left \{0, 0.5, 1 \right \} \ or \ \left \{0, 1 \right \}
+\f]
+
+A detailed explanation about these notations can be found in
+
+[1]. Chris Burges, Tal Shaked, Erin Renshaw, et al. Learning to
+     Rank useing Gradient Descent.
+)DOC");
+  }
+};
+
+class RankLossGradOp : public framework::OperatorWithKernel {
+ public:
+  RankLossGradOp(const std::string &type,
+                 const framework::VariableNameMap &inputs,
+                 const framework::VariableNameMap &outputs,
+                 const framework::AttributeMap &attrs)
+      : OperatorWithKernel(type, inputs, outputs, attrs) {}
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("P"), "Input(P) shouldn't be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Oi"), "Input(Oi) shouldn't be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Oj"), "Input(Oj) shouldn't be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar(framework::GradVarName("Out")),
+                            "Input(Out@GRAD) shouldn't be null.");
+    auto dims = ctx.Input<framework::Tensor>("P")->dims();
+    ctx.Output<framework::Tensor>(framework::GradVarName("P"))->Resize(dims);
+    ctx.Output<framework::Tensor>(framework::GradVarName("Oi"))->Resize(dims);
+    ctx.Output<framework::Tensor>(framework::GradVarName("Oj"))->Resize(dims);
```

Comment:
Gradient Op's output(the gradient of forward op's inputs) can be `nullptr`, which means they are not necessary for backward. So we shall assert it is not `nullptr` before `Resize`.

See https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/cos_sim_op.cc#L146

Reply:
Done